tests/embedding/test_app.py

```diff
@@ -323,3 +327,111 @@ def test_app_execute_script(embedded_app):
+            self.ops.append("get_saved_session")
+
+            # Get name of NamedTemporaryFile
+            self.ops.append("get_name_temp_file")
+
+            # Save app with name of temporary file
+            self.ops.append("save_as")
+
+            return "", ""
+
+        def open_original(self, app, mechdb_file):
+            """Open the original mechdb file from save_original()."""
+            self.ops.append("open_orig_mechdb")
+
+        def graphically_launch_temp(self, app, mechdb_file, temp_file):
+            """Launch the GUI for the mechdb file with a temporary name from save_temp_copy()."""
+            self.ops.append("launch_temp_mechdb")
+
+            return []
+
+    # Create an instance of the MockLauncher object
+    m = MockLauncher()
+
+    # Check that _launch_ui raises an Exception when it hasn't been saved yet
+    with pytest.raises(Exception):
+        _launch_ui(embedded_app, m)
+
+    mechdb_path = os.path.join(tmp_path, "test.mechdb")
+    # Save a test.mechdb file
+    embedded_app.save(mechdb_path)
+    # Launch the UI with the mock class
+    _launch_ui(embedded_app, False, m)
+    # Close the embedded_app
+    embedded_app.close()
+
+    assert m.ops[0] == "save"
+    assert m.ops[1] == "get_saved_session"
+    assert m.ops[2] == "get_name_temp_file"
+    assert m.ops[3] == "save_as"
+    assert m.ops[4] == "open_orig_mechdb"
+    assert m.ops[5] == "launch_temp_mechdb"
+
+
+@pytest.mark.embedding
+def test_launch_gui(embedded_app, tmp_path: pytest.TempPathFactory, capfd):
+    """Test the GUI is launched for an embedded app."""
+    mechdb_path = os.path.join(tmp_path, "test.mechdb")
+    embedded_app.save(mechdb_path)
+    embedded_app.launch_gui(delete_tmp_on_close=False)
+    embedded_app.close()
+    out, err = capfd.readouterr()
+    assert f"Opened a new mechanical session based on {mechdb_path}" in out
+
+
+@pytest.mark.embedding
+def test_launch_gui_exception(embedded_app, tmp_path: pytest.TempPathFactory, capfd):
+    """Test an exception is raised when the embedded_app has not been saved yet."""
+    # Assert that an exception is raised
+    with pytest.raises(Exception):
+        embedded_app.launch_gui(delete_tmp_on_close=False)
+    embedded_app.close()
+
+
+@pytest.mark.embedding_scripts
+def test_tempfile_cleanup(tmp_path: pytest.TempPathFactory, run_subprocess):
+    """Test cleanup function to remove the temporary mechdb file and folder."""
+    temp_file = tmp_path / "tempfiletest.mechdb"
+    temp_folder = tmp_path / "tempfiletest_Mech_Files"
+
+    # Make temporary file
+    temp_file.touch()
+    # Make temporary folder
+    temp_folder.mkdir()
+
+    # Assert the file and folder exist
+    assert temp_file.exists()
+    assert temp_folder.exists()
+
+    # Run process
+    process, stdout, stderr = run_subprocess(["sleep", "3"])
+
+    # Remove the temporary file and folder
+    sys.argv[1:] = [process.pid, temp_file]
```

Comment:
don't modify the process wide sys.argv - that could mess up pytest.  Run the `cleanup_gui_main` script as a subprocess here.